core/src/main/scala/com/databricks/labs/remorph/preprocessor/README.md

```diff
@@ -0,0 +1,205 @@
+   we will not be able to process some esoteric uses of templates, which are generally abuses of the concept.
+   For instance `SELECT A {{ , }} B {{ , }} C` is not a reasonable construct. It would need to be dealt
+   with manually.
+
+### In-place parsing
+
+While similar to the above, we could capture and manage the actual templates in the dialect grammars. This
+removes the need for a pre-processor but means that we woudl need to include more pieces of ANTLR
+specification in each dialect grammar. It also smells like not separating clear responsibilities/functionality.
+
+#### Advantages
+
+ - No need for a pre-processor as such
+
+#### Disadvantages
+
+ - We end up with code dealing with templates intermingled with the dialect parsing code
+ - We still need to track the templates and their locations, so we are not saving any coding effort
+
+### Translate on the fly
+
+One other approach would be to defer the translation of the templates until the SQL is produced by
+DBT, then convert on the fly. 
+
+#### Advantages
+
+ - No need to modify the dialect grammars
+ - No need to track templates
+
+#### Disadvantages
+
+ - We do not guarantee that we can perform a 100% conversion of the incoming source dialect
+ - Users would then be maintaining code in the source dialect, which is not the goal of the project
+
+## The preprocessor
+
+We therefore conclude that the best approach is to use a preprocessor that will locate and track the templates
+and replace them in the pre-processed stream with a placeholder.
+
+As the templates are just text with the macro types above sprinkled in, we can create a preprocessor
+that will always run against the given input text, even if, with un-templated SQL, the preprocessor
+will merely pass through what it sees. However, there may well be other functionality for the preprocessor
+to provide in the future such as perhaps parameter tracking/processing, or perhaps daisy-chaining of
+pre-processors that perform specific tasks.
+
+Hence the preprocessor will be a simple text processor that will:
+
+ - find and replace all `{{ }}`, `{# #}` and `{% %}` blocks with a placeholder
+ - track the unique placeholders and their original text
+ - pass through the placeholders to the target SQL lexer
+ - modify the grammars such that the placeholders are accepted at strategic points in the parse
+ - attempt to convert any SQL within expression macros (complicated by the fact that the Jinga 
+   syntax contains things that look like function calls - we are likely to need a Jinga parser
+ - attempt to convert the SQL within the text where the macros are replaced with placeholders.
+ - the code generator will just pass the placeholders back to the template processor and 
+   the template processor will replace the placeholders with the translated templates.
+
+## Workflow
+```mermaid
+---
+title: Preprocessor Workflow
+config:
+  direction: RL
+  theme: dark
+---
+flowchart TD
```

Comment:
can you adjust this diagram to use our existing concepts - `SnowflakeParser` / `TSqlParser`, `PlanParser`, `LogicalPlanGenerator`, `ExpressionGenerator`, `LogicalPlan`, `Expression` and `ParserRuleContext` + `Parser` from antlr? i struggle to see what components are existing and what components are added here.